### PR TITLE
GetProofByHash: Don't include leaves that aren't in the requested TreeSize

### LINKED ISSUE #1152: Server Skew?
WaitForInclusion(): rpc error: code = InvalidArgument desc = invalid parameter for inclusion proof: index 773 is >= snapshot 773

The pseudocode for the WaitForInclusion function is:
- Fetch current STH
- GetInclusionProofByHash with that STH as the reference STH. 

Is this a reasonable method for testing inclusion, or is it too susceptible to server skew?  

It appears that the particular item we were requesting got sequenced in between these two requests.

### PR DESCRIPTION
**Pull request by gdbelvin**
`GetProofByHash` should not return leaves that >= `req.TreeSize`. 
If no leaves have an index `< req.TreeSize`, `GetProofByHash` should return `codes.NotFound`. 

This was triggering an edge condition when a client:

* Got the current STHA
* Queued a leaf
* *The server then sequenced the leaf*
* Requested a ProofByHash for the leaf using the previous STHA

The resulting proof fails to verify against STHA because the leaf has an index >= STHA.TreeSize.

Fixes #1152

### CODE REVIEW

File: server/log_rpc_server.go [hunk 1]
```diff
@@ -324,9 +319,16 @@ func (t *TrillianLogRPCServer) GetInclusionProofByHash(ctx context.Context, req
...
 		return nil, err
 	}
+	if len(proofs) < 1 {
+		return nil, status.Errorf(codes.NotFound,
+			"No leaf found for hash: %x in tree size %v", req.LeafHash, req.TreeSize)
+	}
 
-	r.Proof = proofs
-	return r, nil
+	// TODO(gbelvin): Rename "Proof" -> "Proofs"
```
Comment (daviddrysdale) [inline server/log_rpc_server.go]: I think `Proof` is correct -- there's a single proof of inclusion for a single leaf in the API response.  

(What's a bit confusing is that this function only deals with a single leaf, but is written to deal with multiple leaves; i.e. the local variable `proofs` has an overly general name).

Comment (gdbelvin) [reply server/log_rpc_server.go]: `Proof` is also an a repeated proof object.  Is that intentional?

Comment (daviddrysdale) [reply server/log_rpc_server.go]: Ah, I missed the [helpful comment in the API](https://github.com/google/trillian/blob/master/trillian_log_api.proto#L202-L205) that one leaf could get multiple inclusion proofs if the underlying tree allows dups (which I don't think is possible anymore?).

It does have a `TODO(gbelvin) only return one proof.` attached to it which kind of contradicts the TODO you're adding here though .... :-)

Comment (gdbelvin) [reply server/log_rpc_server.go]: Hmm, sounds like `TODO(gbelvin) only return one proof` should be removed?

**Comment by daviddrysdale** [comment]
Looks like there's something in the `client/` tests that maybe relies on the previous behaviour?

**Comment by gdbelvin** [comment]
The breadcrumb `fmt.Errorf` was hiding the `status.Errorf` response code.

**Pull request merged**